---
title: "Evaluationg LoopSeq 16S on Rinsate from Retail Meat Samples"
author: "BJC"
output: html_document
---

## Setup

Load libraries, set the working directory to the location of this Rmarkdown file (only necessary when running by hand), and read in filenames:
```{r, echo=FALSE}
library(dada2, quietly=TRUE); packageVersion("dada2") # Should be 1.15.4 or later
library(Biostrings, quietly=TRUE)
library(ShortRead, quietly=TRUE)
library(DECIPHER, quietly=TRUE)
library(ggmsa, quietly=TRUE) # On CRAN, also needs BioC treeio package
library(ggplot2, quietly=TRUE)
library(reshape2, quietly=TRUE)
set.seed(100)
setwd("~/LoopManuscript") # CHANGE ME to the location of this file
path <- "~/LoopData/16S/CallahanMeat" # CHANGE ME to the location of the fastq files
path.fig <- "Figures" # Relative path to where figures should be saved
path.rds <- "RDS" # Relative path to where RDS should be saved
fn <- list.files(path, pattern=".fq$", full.names=TRUE)
sapply(fn, function(f) length(getSequences(f)))
```

## QA, Filtering and Trimming

First do basic QA on these data, and then apply the DADA2 filtering and trimming workflow with the parameters selected from the detailed inspection of the Zymo mock community. Note that this data was generated alongside the Fecal data, so all parameters used there are expected to translate.

```{r}
plotComplexity(fn)
```

```{r}
plotQualityProfile(fn)
```

Remove the primers (and any flanking sequence) from the reads, and filter out reads that don't contain both primers:
```{r}
FWD <- "AGAGTTTGATCMTGGC" # Loop 16S forward primer
REV <- "TACCTTGTTACGACTT" # Loop 16S reverse primer
nop <- file.path(path, "nop", basename(fn))
out <- removePrimers(fn, nop, FWD, rc(REV), verbose=TRUE)
```

As in the fecal samples, only a modest (~40\%) proportion of sequences are passing the primer screen.

Filter the sequences and enforce minimum/maximum lengths appropriate for full-length 16S. Note that we are enforcing `maxEE=1`, as that was determined to be a better filter than `maxEE=2` in the Zymo mock community data.
```{r}
filt <- file.path(path, "filtered", basename(fn))
track <- filterAndTrim(nop, filt, maxEE=1, minLen=1400, maxLen=1600, verbose=TRUE)
```

Final inspection of the quality profile:
```{r, message=FALSE}
plotQualityProfile(filt)
```

Final progress of reads through full-length primer enforcement and filtering/trimming:
```{r}
cbind(raw=out[,1], full.len=out[,2], filtered=track[,2])
```

Few reads lost to quality filtering. Note that GB8 in particular had a large fraction (~80\%) of reads lost due to a failure to detect both primers.

# Denoising

Learn the error rates:
```{r, warning=FALSE}
err.rds <- file.path(path.rds, "err_16S_Meat.rds") # RDS save/load to speed up reruns of the code
if(!file.exists(err.rds)) {
  err <- learnErrors(filt, multi=TRUE, verbose=0)
  saveRDS(err, err.rds)
}
err <- readRDS(err.rds)
plotErrors(err, nominalQ=TRUE)
```

Denoise the filtered data into ASVs using current LoopSeq appropriate "high-sensitivity" settings:
```{r}
dd <- dada(filt, err, OMEGA_A=1e-10, DETECT_SINGLETONS=TRUE, multi=TRUE, verbose=0)
dd
```

Make sequence table and remove chimeras:
```{r}
sta <- makeSequenceTable(dd)
st <- removeBimeraDenovo(sta, minFoldParentOverAbundance=4.5, multi=TRUE, verbose=TRUE)
```

Assign taxonomy:
```{r}
tax.rds <- file.path(path.rds, "tax_16S_Meat.rds") # RDS save/load to speed up reruns of the code
if(!file.exists(tax.rds)) {
  tax <- assignTaxonomy(st, "~/tax/silva_nr_v132_train_set.fa.gz", minBoot=80, multi=TRUE)
  saveRDS(tax, tax.rds)
}
tax <- readRDS(tax.rds)
if(!identical(getSequences(tax), getSequences(st))) stop("Taxonomy mismatch.")
```

## Species-level assignment of CDC foodborne pathogens

Here we are going to focus on 6 species of foodborne pathogens that are known to commonly occur in retail meat according to the CDC:

*Campylobacter: https://www.cdc.gov/campylobacter/faq.html
*Clostridium perfringens: https://www.cdc.gov/foodsafety/diseases/clostridium-perfringens.html (some strains)
*E. coli: https://www.cdc.gov/ecoli/ (STEC, EHEC, ETEC, Diarrheagenic types)
*Listeria monocytogenes: https://www.cdc.gov/listeria/index.html
*Salmonella enterica: https://www.cdc.gov/salmonella/ (some serotypes, esp Enteritidis, Typhimurium, Newport, and Javiana)
*Yersinia enterocolitica: https://www.cdc.gov/yersinia/

First we'll look for the presence of these bacterial genera in our samples (which to this point have only been assigned taxonomy down to the genus level).

```{r}
genera.path <- c("Campylobacter",
                 "Clostridium_sensu_stricto_1", # The Silva 132 genus of C. perfringens
                 "Escherichia/Shigella",
                 "Listeria",
                 "Salmonella",
                 "Yersinia")
genera.path %in% tax[,"Genus"]
```

*Campylobacter* and *Listeria* were not detected in our data. We would like to make species-level assignments for the four pathogenic genera that were in our data, to determine if the pathogen species is there and not a congeneric. This will be done by hand via manual inspection of BLAST searches of the relevant ASVs, so as to avoid any questions about how effectively current species-level assignment tools perform.

### Yersinia (enterocolitica)
```{r}
sq.yers <- getSequences(st)[tax[,"Genus"] %in% "Yersinia"]
names(sq.yers) <- paste0("Yersinia", seq_along(sq.yers))
unname(colSums(st[,sq.yers]))
```

The abundance of Yersinia ASVs drops off quickly. For convenience we'll just do by-hand species assignments for the top 32 (i.e. all Yersinia ASVs with at least an abundance of 4 across all samples).
```{r}
## dada2:::pfasta(sq.yers[1:32])
## BLAST against nt
spec.yers <- c("enterocolitica", # 100
               "enterocolitica",
               "enterocolitica",
               "intermedia", # 100
               "intermedia", # 100
               "intermedia",
               "intermedia",
               "intermedia",
               "kristensenii", # 100
               "intermedia", # 100  -- 10th seq
               "intermedia", # 100
               "intermedia",
               "intermedia",
               "intermedia", #100
               "intermedia",
               "intermedia",
               "intermedia",
               "intermedia", #100
               "enterocolitica", # 100
               "frederiksenii", # -- 20th seq
               "enterocolitica",
               "enterocolitica",
               "enterocolitica",
               "frederiksenii", # 10-off
               "intermedia",
               "enterocolitica", # end of query not included (5% of read)
               "kristensenii",
               "enterocolitica",
               "intermedia", # 8-off of intermedia/kristensenii/aldovae, but cooccur patterns say intermedia
               "enterocolitica", # -- 30th seq
               "enterocolitica",
               "frederiksenii")
names(spec.yers) <- sq.yers[1:length(spec.yers)]
```

All Yersinia ASVs were able to be relatively unambiguously assigned to a specific species, with the arguable exception of Yerisinia29, which was resolved by looking at the co-occurence patterns between that ASV and other unambiguously assigned Yerisina ASVs.

*Y. enterocolitica* is the main foodborne pathogen of concern here, the other Yersinias are not or very rarely associated with disease in humans.

### Salmonella

```{r}
sq.salm <- getSequences(tax)[tax[,"Genus"] %in% "Salmonella"]
uncolname <- function(x) { y <- x; colnames(y) <- NULL; y }
uncolname(st[,sq.salm])
```

Interesting, *Salmonella* only shows up in one sample, and in very even counts across 7 ASVs. That is consistent with the 7 rrn copies that *Salmonella* has according to rrndb. So, what is that strain?

```{r}
## dada2:::pfasta(sq.salm)
## BLAST against nt
spec.salm <- c("enterica",
               "enterica",
               "enterica",
               "enterica",
               "enterica",
               "enterica",
               "enterica")
names(spec.salm) <- sq.salm
```

All Salmonella ASVs are unambiguously *S. enterica* subsp. *enterica*.

Within that (sub-)species, some ASVs (e.g. 1 and 2) are not very discriminative, but Salmonella enterica subsp. enterica serovar Newport strain CFSAN003387, complete genome (https://www.ncbi.nlm.nih.gov/nucleotide/CP016014.1?report=genbank&log$=nucltop&blast_rank=37&RID=9CCPUEXF014) is the best match (usually tied with many others) for every single ASV. All are exact matches except ASV4 (2-off). ASV6 is the key intra-species disciminator, as this Newport accession is the only exact match to that ASV.

### Clostridium (perfringens)

```{r}
sq.Css1 <- getSequences(st)[tax[,"Genus"] %in% "Clostridium_sensu_stricto_1"]
uncolname(st[,sq.Css1])
```

This genus only shows up in sample GT10 with any frequency, but also shows up at low numbers in CB28 and GT8.

```{r}
## dada2:::pfasta(sq.Css1) # C. perfringens group
## BLAST against nt
spec.Css1 <- c("perfringens", # 100
"perfringens",
"perfringens",
"perfringens", # 100
"perfringens",
"perfringens",
"perfringens",
"perfringens",
"perfringens",
"septicum", # SEQ 10
"perfringens",
"perfringens",
"septicum",
"septicum",
"perfringens",
"perfringens",
"perfringens", # 23-off
"perfringens",
"perfringens",
"perfringens", # SEQ 20
"septicum",
"perfringens")
names(spec.Css1) <- sq.Css1
```

Most of the ASVs are unambiguously assigned to *C. perfringens*, but some are clearly the related *C. septicum* (not a pathogen concern).

### Escherichia/Shigella coli

```{r}
sq.ec<- getSequences(tax)[tax[,"Genus"] %in% "Escherichia/Shigella"]
unname(colSums(st[,sq.ec]))
```

The abundance of Escherichia/Shigella ASVs drops off quickly. Using the same standard as before, we'll take a look at the first 68 ASVs, representing all ASVs with at least 4 reads study-wide.

```{r}
## dada2:::pfasta(sq.ec[1:68])
## BLAST against nt
# First 60 all e. coli except 10/39/41/42/51/54/55/68, Shigella also i top hits
# 66: E.coli and E. fergusonii in top hits
```

The results here require some quick unpacking. First, the vast majority of these ASVs have *E. coli* and only *E. coli* among their top hits and can be unambiguously assigned to the *E. coli* species. Sequences 10, 39, 41, 42, 51, 54, 55 and 68 have both *E. coli* and some species of *Shigella* among their top hits. However, given the (in)famous status of *Shigella* as a parphyletic genus entirely contained within the *E. coli* species, and that this well-known inconsistency only remains due to the historical and clinical usage of those names, we simply consider these as *E. coli* here. Finally, a single sequences 66 has both *E. coli* and *E. fergusonii* in its top hits, thus cannot be unambiguously assigned to the species level.

```{r}
spec.ec <- rep("coli", 68)
spec.ec[66] <- NA
names(spec.ec) <- sq.ec[1:68]
```

```{r}
uncolname(st[,sq.ec[1:68]])
```

Essentially no *E. coli* in the first 3 samples. *E. coli* abundance is high in GT10 and GT8, however, with the numbers of ASVs present in each suggesting multiple *E. coli* strains present in each sample (*E. coli* has 7 copies of the 16S rRNA gene, so at most 7 alleles per strain). 

## Figures

First let's plot the relative abundances and identities of the CDC key foodborne pathogens in these 6 samples: Yersinia enterolitica, E. coli (Escherichia/Shigella), Salmonella enterica, Clostridium perfringens (Clostridium_sensu_stricto_1), Campylobacter and Listeria monocytogenes.
```{r}
tt <- tax
tt <- cbind(tt, Species=NA)
tt[names(spec.yers),"Species"] <- spec.yers
tt[names(spec.salm),"Species"] <- spec.salm
tt[names(spec.Css1),"Species"] <- spec.Css1
tt[names(spec.ec),"Species"] <- spec.ec
table(tt[,"Species"], useNA="ifany")
```

OK, so we have our taxonomy table with species defined for all the genera containing the 6 species of interest, only 4 of which are present. Let's make a relative abundance barplot showing their relative abundances of these CDC species-of-interest in these samples:
```{r}
ft <- sweep(st, 1, rowSums(st), "/")
spec.nice <- c("enterocolitica"="Y. enterocolitica", 
               "perfringens"="C. perfringens", 
               "enterica"="S. enterica", 
               "coli"="E. coli", 
               "monocytogenes"="L. monocytogenes", 
               "Campylobacter"="Campylobacter sp.")
tt[!tt[,"Species"] %in% names(spec.nice),"Species"] <- NA # Drop non-pathogen species
tt[!is.na(tt[,"Species"]), "Species"] <- spec.nice[tt[!is.na(tt[,"Species"]), "Species"]]
suppressWarnings(ftp <- t(rowsum(t(ft), group=tt[,"Species"]))) 
# Throws a warning because of the NAs in species, this is OK
ftp <- ftp[,colnames(ftp) %in% spec.nice]
ftp
```

```{r}
dfp <- cbind(data.frame(Sample=substr(rownames(ftp), 1, 3), stringsAsFactors=FALSE), 
             data.frame(ftp, stringsAsFactors=FALSE))
dfpm <- melt(dfp, id.vars="Sample", variable.name="Species", value.name="Proportion") ### WARNING: Auto-converts the variable name to a FACTOR
dfpm$Species <- as.character(dfpm$Species)
dfpm$Species <- gsub("[.]{2}", ". ", dfpm$Species)
dfpm$Species <- factor(dfpm$Species, levels=spec.nice)
```

```{r}
fill.scale.path <- scale_fill_manual(values=c("C. perfringens"="#1b9e77", "E. coli"="#d95f02", "S. enterica"="#7570b3", 
                                              "Y. enterocolitica"="#e7298a", "L. monocytogenes"="tomato4", "Campylobacter sp."="#e6ab02"),
                                     name="Pathogenic\nspecies", drop=FALSE)
ppp <- ggplot(data=dfpm, aes(x=Sample, y=Proportion, fill=Species)) + geom_col(width=0.4) + ylim(0,1) + 
  fill.scale.path + theme_bw() + theme(panel.grid=element_blank()) + theme(legend.text=element_text(face="italic"))
ppp
ggsave(file.path(path.fig, "Meat_PathogenProportions.pdf"), ppp, width=6, height=2.5, units="in", useDingbats=FALSE)
ggsave(file.path(path.fig, "Meat_PathogenProportions.png"), ppp, width=6, height=2.5, units="in")
```

Let's return to *C. perfringens* as an example of high-res full complement "ribotyping" that is possible with this method. Recall the distribution of *C. perfringens* ASVs:
```{r}
sq.Cperf <- sq.Css1[spec.Css1 %in% "perfringens"]
uncolname(st[,sq.Cperf])
```

We are interested ine visualizing the "ribotype" of the strain in Sample GT10 that spans *C. perfringens* ASVs 1-9, with a 2:1(x8) ratio of alleles consistent with the 10 copies of the 16S rRNA gene in *C. perfringens* genomes.

```{r}
sq.Cpstrain <- sq.Cperf[1:9]
dna.Cpstrain <- DNAStringSet(sq.Cpstrain)
rna.Cpstrain <- DNAStringSet(dna.Cpstrain)
aln <- AlignSeqs(rna.Cpstrain) # align with RNA secondary structure
names(aln) <- paste0("Cperf", seq_along(sq.Cpstrain))
aln
```

Scannings 100nt chunks to see where some diffs are for the publication plot:
```{r, warnings=FALSE, message=FALSE}
lst <- min(nchar(aln))
for(start in seq(1, 1500, 100)) {
  print(ggmsa(aln, start, min(lst, start+99), font = NULL, color = 'Shapely_NT'))
}
```

Almost all the diffs are in V1/V2 it seems (about postiions 50-75 and 150-190 in these sequences, which start at ~E. coli coordinate position 27, which overlap the V1 and V2 E. coli positions according to https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2562909/).

Maybe just plot V2 using those coords:

```{r}
pmsa <- ggmsa(aln, 137-27, 242-27, font = NULL, color = 'Shapely_NT') + 
  ggtitle(expression(italic("C. perfringens")~"alleles, V2 hypervariable region")) +
  theme(axis.text.y=element_blank()) + theme(plot.title = element_text(hjust = 0.5))
pmsa
ggsave(file.path(path.fig, "Meat_CPerf_MSA.pdf"), pmsa, width=6, height=1.2, units="in", useDingbats=FALSE)
ggsave(file.path(path.fig, "Meat_CPerf_MSA.png"), pmsa, width=6, height=1.2, units="in")
```

Add the abundances as a column to the right:
```{r}
df.ab <- data.frame(Name=factor(names(aln), levels=sort(names(aln), decreasing=TRUE)), 
                    Sequence=sq.Cpstrain, 
                    Abundance=st["GT10_contig_list_trimmed.fq", sq.Cpstrain])
pab <- ggplot(df.ab, aes(x=Abundance, y=Name)) +
  geom_col() +
  theme_minimal() +
  theme(panel.grid=element_blank()) +
  theme(axis.text=element_blank()) +
  ylab("") + xlab("Relative\nAbundance")
pab
ggsave(file.path(path.fig, "Meat_CPerf_MSA_ab.pdf"), pab, width=1, height=1.2, units="in", useDingbats=FALSE)
ggsave(file.path(path.fig, "Meat_CPerf_MSA_ab.png"), pab, width=1, height=1.2, units="in")
```


