---
title: "Evaluationg LoopSeq 16S on Rinsate from Retail Meat Samples"
author: "BJC"
date: "4/13/2020"
output: html_document
---

## Setup

Load libraries, set the working directory to the location of this Rmarkdown file (only necessary when running by hand), and read in filenames:
```{r, echo=FALSE}
library(dada2, quietly=TRUE); packageVersion("dada2") # Should be 1.15.4 or later
library(Biostrings, quietly=TRUE)
library(ShortRead, quietly=TRUE)
library(ggplot2, quietly=TRUE)
library(reshape2, quietly=TRUE)
set.seed(100)
setwd("~/LoopManuscript") # CHANGE ME to the location of this file
path <- "~/LoopData/16S/CallahanMeat" # CHANGE ME to the location of the fastq files
path.fig <- "Figures" # Relative path to where figures should be saved
path.rds <- "RDS" # Relative path to where RDS should be saved
fn <- list.files(path, pattern=".fq$", full.names=TRUE)
sapply(fn, function(f) length(getSequences(f)))
```

## QA, Filtering and Trimming

First do basic QA on these data, and then apply the DADA2 filtering and trimming workflow with the parameters selected from the detailed inspection of the Zymo mock community. Note that this data was generated alongside the Fecal data, so all parameters used there are expected to translate.

```{r}
plotComplexity(fn)
```

```{r}
plotQualityProfile(fn)
```

Remove the primers (and any flanking sequence) from the reads, and filter out reads that don't contain both primers:
```{r}
FWD <- "AGAGTTTGATCMTGGC" # Loop 16S forward primer
REV <- "TACCTTGTTACGACTT" # Loop 16S reverse primer
nop <- file.path(path, "nop", basename(fn))
out <- removePrimers(fn, nop, FWD, rc(REV), verbose=TRUE)
```

As in the fecal samples, only a modest (~40\%) proportion of sequences are passing the primer screen.

Filter the sequences and enforce minimum/maximum lengths appropriate for full-length 16S. Note that we are enforcing `maxEE=1`, as that was determined to be a better filter than `maxEE=2` in the Zymo mock community data.
```{r}
filt <- file.path(path, "filtered", basename(fn))
track <- filterAndTrim(nop, filt, maxEE=1, minLen=1400, maxLen=1600, verbose=TRUE)
```

Final inspection of the quality profile:
```{r, message=FALSE}
plotQualityProfile(filt)
```

Final progress of reads through full-length primer enforcement and filtering/trimming:
```{r}
cbind(raw=out[,1], full.len=out[,2], filtered=track[,2])
```

Few reads lost to quality filtering. Note that GB8 in particular had a large fraction (~80\%) of reads lost due to a failure to detect both primers.

# Denoising

Learn the error rates:
```{r, warning=FALSE}
err.rds <- file.path(path.rds, "err_16S_Meat.rds") # RDS save/load to speed up reruns of the code
if(!file.exists(err.rds)) {
  err <- learnErrors(filt, multi=TRUE, verbose=0)
  saveRDS(err, err.rds)
}
err <- readRDS(err.rds)
plotErrors(err, nominalQ=TRUE)
```

Denoise the filtered data into ASVs using current LoopSeq appropriate "high-sensitivity" settings:
```{r}
dd <- dada(filt, err, OMEGA_A=1e-10, DETECT_SINGLETONS=TRUE, multi=TRUE, verbose=0)
dd
```

Make sequence table and remove chimeras:
```{r}
sta <- makeSequenceTable(dd)
st <- removeBimeraDenovo(sta, minFoldParentOverAbundance=4.5, multi=TRUE, verbose=TRUE)
```

Assign taxonomy:
```{r}
tax.rds <- file.path(path.rds, "tax_16S_Meat.rds") # RDS save/load to speed up reruns of the code
if(!file.exists(tax.rds)) {
  tax <- assignTaxonomy(st, "~/tax/silva_nr_v132_train_set.fa.gz", minBoot=80, multi=TRUE)
  saveRDS(tax, tax.rds)
}
tax <- readRDS(tax.rds)
if(!identical(getSequences(tax), getSequences(st))) stop("Taxonomy mismatch.")
```

## Species-level assignment of CDC foodborne pathogens

Here we are going to focus on 6 species of foodborne pathogens that are known to commonly occur in retail meat according to the CDC:

*Campylobacter: https://www.cdc.gov/campylobacter/faq.html
*Clostridium perfringens: https://www.cdc.gov/foodsafety/diseases/clostridium-perfringens.html (some strains)
*E. coli: https://www.cdc.gov/ecoli/ (STEC, EHEC, ETEC, Diarrheagenic types)
*Listeria monocytogenes: https://www.cdc.gov/listeria/index.html
*Salmonella enterica: https://www.cdc.gov/salmonella/ (some serotypes, esp Enteritidis, Typhimurium, Newport, and Javiana)
*Yersinia enterocolitica: https://www.cdc.gov/yersinia/

First we'll look for the presence of these bacterial genera in our samples (which to this point have only been assigned taxonomy down to the genus level).

```{r}
genera.path <- c("Campylobacter",
                 "Clostridium_sensu_stricto_1", # The Silva 132 genus of C. perfringens
                 "Escherichia/Shigella",
                 "Listeria",
                 "Salmonella",
                 "Yersinia")
genera.path %in% tax[,"Genus"]
```

*Campylobacter* and *Listeria* were not detected in our data. We would like to make species-level assignments for the four pathogenic genera that were in our data, to determine if the pathogen species is there and not a congeneric. This will be done by hand via manual inspection of BLAST searches of the relevant ASVs, so as to avoid any questions about how effectively current species-level assignment tools perform.

### Yersinia (enterocolitica)
```{r}
sq.yers <- getSequences(st)[tax[,"Genus"] %in% "Yersinia"]
names(sq.yers) <- paste0("Yersinia", seq_along(sq.yers))
unname(colSums(st[,sq.yers]))
```

The abundance of Yersinia ASVs drops off quickly. For convenience we'll just do by-hand species assignments for the top 32 (i.e. all Yersinia ASVs with at least an abundance of 4 across all samples).
```{r}
## dada2:::pfasta(sq.yers[1:32])
## BLAST against nt
spec.yers <- c("enterocolitica", # 100
               "enterocolitica",
               "enterocolitica",
               "intermedia", # 100
               "intermedia", # 100
               "intermedia",
               "intermedia",
               "intermedia",
               "kristensenii", # 100
               "intermedia", # 100  -- 10th seq
               "intermedia", # 100
               "intermedia",
               "intermedia",
               "intermedia", #100
               "intermedia",
               "intermedia",
               "intermedia",
               "intermedia", #100
               "enterocolitica", # 100
               "frederiksenii", # -- 20th seq
               "enterocolitica",
               "enterocolitica",
               "enterocolitica",
               "frederiksenii", # 10-off
               "intermedia",
               "enterocolitica", # end of query not included (5% of read)
               "kristensenii",
               "enterocolitica",
               "intermedia", # 8-off of intermedia/kristensenii/aldovae, but cooccur patterns say intermedia
               "enterocolitica", # -- 30th seq
               "enterocolitica",
               "frederiksenii")
names(spec.yers) <- sq.yers[1:length(spec.yers)]
```

All Yersinia ASVs were able to be relatively unambiguously assigned to a specific species, with the arguable exception of Yerisinia29, which was resolved by looking at the co-occurence patterns between that ASV and other unambiguously assigned Yerisina ASVs.

*Y. enterocolitica* is the main foodborne pathogen of concern here, the other Yersinias are not or very rarely associated with disease in humans.

### Salmonella

```{r}
sq.salm <- getSequences(tax)[tax[,"Genus"] %in% "Salmonella"]
uncolname <- function(x) { y <- x; colnames(y) <- NULL; y }
uncolname(st[,sq.salm])
```

Interesting, *Salmonella* only shows up in one sample, and in very even counts across 7 ASVs. That is consistent with the 7 rrn copies that *Salmonella* has according to rrndb. So, what is that strain?

```{r}
## dada2:::pfasta(sq.salm)
## BLAST against nt
spec.salm <- c("enterica",
               "enterica",
               "enterica",
               "enterica",
               "enterica",
               "enterica",
               "enterica")
names(spec.salm) <- sq.salm
```

All Salmonella ASVs are unambiguously *S. enterica* subsp. *enterica*.

Within that (sub-)species, some ASVs (e.g. 1 and 2) are not very discriminative, but Salmonella enterica subsp. enterica serovar Newport strain CFSAN003387, complete genome (https://www.ncbi.nlm.nih.gov/nucleotide/CP016014.1?report=genbank&log$=nucltop&blast_rank=37&RID=9CCPUEXF014) is the best match (usually tied with many others) for every single ASV. All are exact matches except ASV4 (2-off). ASV6 is the key intra-species disciminator, as this Newport accession is the only exact match to that ASV.

### Clostridium (perfringens)

```{r}
sq.Css1 <- getSequences(st)[tax[,"Genus"] %in% "Clostridium_sensu_stricto_1"]
uncolname(st[,sq.Css1])
```

This genus only shows up in sample GT10 with any frequency, but also shows up at low numbers in CB28 and GT8.

```{r}
## dada2:::pfasta(sq.Css1) # C. perfringens group
## BLAST against nt
spec.Css1 <- c("perfringens", # 100
"perfringens",
"perfringens",
"perfringens", # 100
"perfringens",
"perfringens",
"perfringens",
"perfringens",
"perfringens",
"septicum", # SEQ 10
"perfringens",
"perfringens",
"septicum",
"septicum",
"perfringens",
"perfringens",
"perfringens", # 23-off
"perfringens",
"perfringens",
"perfringens", # SEQ 20
"septicum",
"perfringens")
names(spec.Css1) <- sq.Css1
```

Most of the ASVs are unambiguously assigned to *C. perfringens*, but some are clearly the related *C. septicum* (not a pathogen concern).

### Escherichia/Shigella coli

```{r}
sq.ec<- getSequences(tax)[tax[,"Genus"] %in% "Escherichia/Shigella"]
unname(colSums(st[,sq.ec]))
```

The abundance of Escherichia/Shigella ASVs drops off quickly. Using the same standard as before, we'll take a look at the first 68 ASVs, representing all ASVs with at least 4 reads study-wide.

```{r}
## dada2:::pfasta(sq.ec[1:68])
## BLAST against nt
# First 60 all e. coli except 10/39/41/42/51/54/55/68, Shigella also i top hits
# 66: E.coli and E. fergusonii in top hits
```

The results here require some quick unpacking. First, the vast majority of these ASVs have *E. coli* and only *E. coli* among their top hits and can be unambiguously assigned to the *E. coli* species. Sequences 10, 39, 41, 42, 51, 54, 55 and 68 have both *E. coli* and some species of *Shigella* among their top hits. However, given the (in)famous status of *Shigella* as a parphyletic genus entirely contained within the *E. coli* species, and that this well-known inconsistency only remains due to the historical and clinical usage of those names, we simply consider these as *E. coli* here. Finally, a single sequences 66 has both *E. coli* and *E. fergusonii* in its top hits, thus cannot be unambiguously assigned to the species level.

```{r}
spec.ec <- rep("coli", 68)
spec.ec[66] <- NA
names(spec.ec) <- sq.ec[1:68]
```

```{r}
uncolname(st[,sq.ec[1:68]])
```

Essentially no *E. coli* in the first 3 samples. *E. coli* abundance is high in GT10 and GT8, however, with the numbers of ASVs present in each suggesting multiple *E. coli* strains present in each sample (*E. coli* has 7 copies of the 16S rRNA gene, so at most 7 alleles per strain). 

## Figures

First let's plot the relative abundances and identities of the CDC key foodborne pathogens in these 6 samples: Yersinia enterolitica, E. coli (Escherichia/Shigella), Salmonella enterica, Clostridium perfringens (Clostridium_sensu_stricto_1), Campylobacter and Listeria monocytogenes.
```{r}
tt <- tax
tt <- cbind(tt, Species=NA)
tt[names(spec.yers),"Species"] <- spec.yers
tt[names(spec.salm),"Species"] <- spec.salm
tt[names(spec.Css1),"Species"] <- spec.Css1
tt[names(spec.ec),"Species"] <- spec.ec
table(tt[,"Species"], useNA="ifany")
```

OK, so we have our taxonomy table with species defined for all the genera containing the 6 species of interest, only 4 of which are present. Let's make a relative abundance barplot showing their relative abundances of these CDC species-of-interest in these samples:
```{r}
ft <- sweep(st, 1, rowSums(st), "/")
spec.nice <- c("enterocolitica"="Y. enterocolitica", 
               "perfringens"="C. perfringens", 
               "enterica"="S. enterica", 
               "coli"="E. coli", 
               "monocytogenes"="L. monocytogenes", 
               "Campylobacter"="Campylobacter sp.")
tt[!tt[,"Species"] %in% names(spec.nice),"Species"] <- NA # Drop non-pathogen species
tt[!is.na(tt[,"Species"]), "Species"] <- spec.nice[tt[!is.na(tt[,"Species"]), "Species"]]
suppressWarnings(ftp <- colsum(ft, group=tt[,"Species"])) # Throws a warning because of the NAs in species, this is OK
ftp <- ftp[,colnames(ftp) %in% spec.nice]
ftp
```

```{r}
dfp <- cbind(data.frame(Sample=substr(rownames(ftp), 1, 3), stringsAsFactors=FALSE), 
             data.frame(ftp, stringsAsFactors=FALSE))
dfpm <- melt(dfp, id.vars="Sample", variable.name="Species", value.name="Proportion") ### WARNING: Auto-converts the variable name to a FACTOR
dfpm$Species <- as.character(dfpm$Species)
dfpm$Species <- gsub("[.]{2}", ". ", dfpm$Species)
dfpm$Species <- factor(dfpm$Species, levels=spec.nice)
```

```{r}
fill.scale.path <- scale_fill_manual(values=c("C. perfringens"="#1b9e77", "E. coli"="#d95f02", "S. enterica"="#7570b3", 
                                              "Y. enterocolitica"="#e7298a", "L. monocytogenes"="tomato4", "Campylobacter sp."="#e6ab02"),
                                     name="Pathogenic\nspecies", drop=FALSE)
ppp <- ggplot(data=dfpm, aes(x=Sample, y=Proportion, fill=Species)) + geom_col(width=0.4) + ylim(0,1) + 
  fill.scale.path + theme_bw() + theme(panel.grid=element_blank()) + theme(legend.text=element_text(face="italic"))
ppp
ggsave(file.path(path.fig, "Meat_PathogenProportions.pdf"), ppp, width=5, height=3, units="in", useDingbats=FALSE)
ggsave(file.path(path.fig, "Meat_PathogenProportions.png"), ppp, width=5, height=3, units="in")
```

Let's return to *C. perfringens* as an example of high-res full complement "ribotyping" that is possible with this method. Recall the distribution of *C. perfringens* ASVs:
```{r}
sq.Cperf <- sq.Css1[spec.Css1 %in% "perfringens"]
uncolname(st[,sq.Cperf])
```

We are interested ine visualizing the "ribotype" of the strain in Sample GT10 that spans *C. perfringens* ASVs 1-9, with a 2:1(x8) ratio of alleles consistent with the 10 copies of the 16S rRNA gene in *C. perfringens* genomes.

```{r}
library(DECIPHER)
library(ggmsa) # On CRAN, also needs BioC treeio package
sq.Cpstrain <- sq.Cperf[1:9]
dna.Cpstrain <- DNAStringSet(sq.Cpstrain)
rna.Cpstrain <- DNAStringSet(dna.Cpstrain)
aln <- AlignSeqs(rna.Cpstrain) # align with RNA secondary structure
names(aln) <- paste0("Cperf", seq_along(sq.Cpstrain))
aln
```

```{r}
## Testing ggmsa nucleotide color schemes:
# clrs <- c('Clustal', 'Chemistry_NT', 'Shapely_NT', 'Zappo_NT', 'Taylor_NT')
# for(clr in clrs) {
#  print(ggmsa(aln, 100, 200, font = NULL, color = clr))
#}
```

Scannings 100nt chunks to see where some diffs are for the publication plot:
```{r, warnings=FALSE, message=FALSE}
lst <- min(nchar(aln))
for(start in seq(1, 1500, 100)) {
  print(ggmsa(aln, start, min(lst, start+99), font = NULL, color = 'Shapely_NT'))
}
```

Almost all the diffs are in V1/V2 it seems (about postiions 50-75 and 150-190 in these sequences, which start at ~E. coli coordinate position 27, which overlap the V1 and V2 E. coli positions according to https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2562909/).

Maybe just plot V2 using those coords:

```{r}
pmsa <- ggmsa(aln, 137-27, 242-27, font = NULL, color = 'Shapely_NT') + 
  ggtitle(expression(italic("C. perfringens")~"alleles, V2 hypervariable region")) +
  theme(axis.text.y=element_blank()) + theme(plot.title = element_text(hjust = 0.5))
pmsa
ggsave(file.path(path.fig, "Meat_CPerf_MSA.pdf"), pmsa, width=8, height=1.2, units="in", useDingbats=FALSE)
ggsave(file.path(path.fig, "Meat_CPerf_MSA.png"), pmsa, width=8, height=1.2, units="in")
```










```{r}
table(tax[,"Phylum"], useNA="ifany")
```

```{r}
table(tax[,"Genus"], useNA="ifany")
```

Quite a lot of Yersinia is interesting, as that can be a foodborne pathogen, or a "commensal" or (infamously in a couple microbiome papers) the plague. What is the abundance spectrum like?

```{r}
sq.yers <- getSequences(st)[tax[,"Genus"] %in% "Yersinia"]
names(sq.yers) <- paste0("Yersinia", seq_along(sq.yers))
unname(colSums(st[,sq.yers]))
```

Drops off quickly, but a number of well-supported Yersinia ASVs (>=10 reads). Taking a look at the Yersinia ASV distribution by sample (and note that rrndb supports mostly 7 gene copies, but 6 in some species as well)

```{r}
df <- cbind(data.frame(Sample=rownames(st)), st[,sq.yers])
dfm <- melt(df, id.vars="Sample", variable.name="Yersinia.ASV", value.name="Abundance")
ggplot(data=dfm[dfm$Abundance > 4,], aes(x=Yersinia.ASV, y=Abundance)) + geom_point() + facet_wrap(~Sample) + theme_bw()
```

Let's do some species-level assignments by hand for the top 32 Yersinias (all w/ 4+ reads).

```{r}
spec.yers <- c("enterocolitica", # 100
               "enterocolitica",
               "enterocolitica",
               "intermedia", # 100
               "intermedia", # 100
               "intermedia",
               "intermedia",
               "intermedia",
               "kristensenii", # 100
               "intermedia", # 100  -- 10th seq
               "intermedia", # 100
               "intermedia",
               "intermedia",
               "intermedia", #100
               "intermedia",
               "intermedia",
               "intermedia",
               "intermedia", #100
               "enterocolitica", # 100
               "frederiksenii", # -- 20th seq
               "enterocolitica",
               "enterocolitica",
               "enterocolitica",
               "frederiksenii", # 10-off
               "intermedia",
               "enterocolitica", # end of query not included (5% of read)
               "kristensenii",
               "enterocolitica",
               "intermedia", # 8-off of intermedia/kristensenii/aldovae, but cooccur patterns say intermedia
               "enterocolitica", # -- 30th seq
               "enterocolitica",
               "frederiksenii")
names(spec.yers) <- sq.yers[1:length(spec.yers)]
```

Y. enterolitica are the most common species causing human enteric (intestinal) yersiniosis. Pigs are the major animal reservoir for the few strains of Y. enterocolitica that cause human illness, but rodents, rabbits, sheep, cattle, horses, dogs, and cats also can carry strains that cause human illness. https://www.cdc.gov/yersinia/faq.html

Yersinia intermedia is not considered of clinical relevance, being isolated from humans in a routine manner. https://en.wikipedia.org/wiki/Yersinia_intermedia

Yersinia kristensenii is a species of bacteria.[1] It is Gram-negative and its type strain is 105 (=CIP 80-30). It is potentially infectious to mice.[2] It secretes a bacteriocin that targets related species. https://en.wikipedia.org/wiki/Yersinia_kristensenii

Yersinia frederiksenii is a Gram-negative species of bacteria.[1] It uses rhamnose and sucrose. Its type strain is strain 6175 (=CIP 80-29). In humans, it can cause gastrointestinal infections,[2] while it has also been found in fish. https://en.wikipedia.org/wiki/Yersinia_frederiksenii

So, *Y. enterolitica* is the main foodborne pathogen of concern here, the other Yersinias are not or very rarely associated with disease in humans.

```{r}
sty <- st[,sq.yers]
dfy <- data.frame(Sample=sapply(strsplit(rownames(st), "_"), `[`, 1), st[,sq.yers])
dfym <- melt(dfy, id.vars="Sample", variable.name="ASV", value.name="Abundance")
dfym$Genus <- "Yersinia"
dfym$Species <- spec.yers[dfym$ASV]
```

```{r}
ggplot(data=dfym[dfym$ASV %in% names(spec.yers) & dfym$Abundance > 0,], 
       aes(x=ASV, y=Abundance, color=Species)) + 
  geom_point() + theme_bw() + facet_grid(Sample~Species, scales="free_y") + ylim(c(0, NA))
```

```{r}
sq.yent <- names(spec.yers)[spec.yers == "enterocolitica"]
unname(outer(sq.yent, sq.yent,
             nwhamming, vec=TRUE))
```

So groups of similar sequences: 1/2/3/4/6/7/11, and 5/8/9 (although these are ~10nts from each other, but 30+ from the rest). 10 is off on its own.

In BLAST the first group are all close to the same Yersinia enterocolitica strain FORC-002, complete genome.
10 is also a match to that but with the end cut off -- chimera?

Actually all of 5/8/9 are that same way...

```{r}
unname(st[,sq.yent])
```

All of 5/8/9/10 cooccur in that first sample, along with the very high abundance top 3 seqs (which are probably one strain).

```{r}
sq.yent.last80 <- substr(sq.yent, nchar(sq.yent)-80, nchar(sq.yent))
dada2:::pfasta(sq.yent.last80)
```

Interesting, the 5/8/9/10 last 80s don't match any Yersinia sequences, and are only poor matches (sometimes) to Serratia in the broader Yersiniaceae group. However they match exactly to Streptococcus/Enterococcus sequences...

Is there a quality signal associated here?

```{r}
plot(dd[[1]]$quality[dd[[1]]$sequence==sq.yent[[1]],])
plot(dd[[1]]$quality[dd[[1]]$sequence==sq.yent[[2]],])
plot(dd[[1]]$quality[dd[[1]]$sequence==sq.yent[[5]],])
plot(dd[[1]]$quality[dd[[1]]$sequence==sq.yent[[10]],])
```

No, not really...

```{r}
bim <- isBimeraDenovo(sta, minFoldParent=4.5, allowOneOff=TRUE, multi=TRUE)
unname(bim[sq.yent])
```

Not sure at this point...

Looking more broadly at the samples using the following, and referencing against https://www.cdc.gov/foodsafety/diseases/index.html 
```{r}
i <- 4
unname(tax[order(st[i,], decreasing=TRUE),"Genus"][1:40])
sort(unname(tax[order(st[i,], decreasing=TRUE),"Genus"][1:40]))
```

It looks like overall I need to look into Escherichia/Shigella (samples 4/6), Clostridium (a few samples) and Yersinia.

List of bacteria of foodborne pathogen concern via CDC:

Bacillus anthracis: https://www.cdc.gov/anthrax/basics/ (prob not retail meat in US at least)
Clostridium botulinum: https://www.cdc.gov/botulism/general.html (prob not retail meat, disease from toxin not the usualy spore bacteria itself)
Brucella: https://www.cdc.gov/brucellosis/index.html (prob not meat, but possible)
*Campylobacter: https://www.cdc.gov/campylobacter/faq.html
Vibrio cholerae: https://www.cdc.gov/cholera/index.html (prob not in meat)
Cronobacter: https://www.cdc.gov/cronobacter/ (prob not in meat)
*Clostridium perfringens: https://www.cdc.gov/foodsafety/diseases/clostridium-perfringens.html (some strains)
*E. coli: https://www.cdc.gov/ecoli/ (STEC, EHEC, ETEC, Diarrheagenic types)
Streptobacillus moniliformis: https://www.cdc.gov/rat-bite-fever/ (prob not in meat)
Leptospira: https://www.cdc.gov/leptospirosis/ (prob not in meat)
*Listeria monocytogenes: https://www.cdc.gov/listeria/index.html
*Salmonella enterica: https://www.cdc.gov/salmonella/ (some serotypes, esp Enteritidis, Typhimurium, Newport, and Javiana)
Salmonella Typhi/Paratyphi: https://www.cdc.gov/typhoid-fever/ (prob not in US retail meat)
Shigella: https://www.cdc.gov/shigella/ 
Staphylococcus aureus: https://www.cdc.gov/foodsafety/diseases/staphylococcal.html (disease from toxin, not the common commensal itself)
Vibrio parahaemolyticus/vulnificus/alginolyticus: https://www.cdc.gov/vibrio/faq.html (prob not in meat)
*Yersinia enterocolitica: https://www.cdc.gov/yersinia/

In our 6 samples, we see quite a bit of Yersinia and Escherichia/Shigella, some Salmonella, and perhaps C. perfringens (not sure yet what sensu_stricto group that is in). No Campylobacter or Listeria is detected.

```{r}
# >NR_121697.2:24-1462 Clostridium perfringens ATCC 13124 16S ribosomal RNA, complete sequence
ex.Cperf <- gsub("\n", "", "TCAGGATGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAGCGATGAAGTTTCCTTCGGGAAACGGAT
TAGCGGCGGACGGGTGAGTAACACGTGGGTAACCTGCCTCATAGAGTGGAATAGCCTTCCGAAAGGAAGA
TTAATACCGCATAACGTTGAAAGATGGCATCATCATTCAACCAAAGGAGCAATCCGCTATGAGATGGACC
CGCGGCGCATTAGCTAGTTGGTGGGGTAACGGCCTACCAAGGCGACGATGCGTAGCCGACCTGAGAGGGT
GATCGGCCACATTGGGACTGAGACACGGCCCAGACTCCTACGGGAGGCAGCAGTGGGGAATATTGCACAA
TGGGGGAAACCCTGATGCAGCAACGCCGCGTGAGTGATGAAGGTTTTCGGATCGTAAAGCTCTGTCTTTG
GGGAAGATAATGACGGTACCCAAGGAGGAAGCCACGGCTAACTACGTGCCAGCAGCCGCGGTAATACGTA
GGTGGCGAGCGTTATCCGGATTTACTGGGCGTAAAGGGAGCGTAGGCGGATGATTAAGTGGGATGTGAAA
TACCCGGGCTCAACTTGGGTGCTGCATTCCAAACTGGTTATCTAGAGTGCAGGAGAGGAGAGTGGAATTC
CTAGTGTAGCGGTGAAATGCGTAGAGATTAGGAAGAACACCAGTGGCGAAGGCGACTCTCTGGACTGTAA
CTGACGCTGAGGCTCGAAAGCGTGGGGAGCAAACAGGATTAGATACCCTGGTAGTCCACGCCGTAAACGA
TGAATACTAGGTGTGGGGGTTTCAACACCTCCGTGCCGCCGCTAACGCATTAAGTATTCCGCCTGGGGAG
TACGGTCGCAAGATTAAAACTCAAAGGAATTGACGGGGACCCGCACAAGTAGCGGAGCATGTGGTTTAAT
TCGAAGCAACGCGAAGAACCTTACCTACACTTGACATCCCTTGCATTACTCTTAATCGAGGAAATCCCTT
CGGGGACAAGGTGACAGGTGGTGCATGGTTGTCGTCAGCTCGTGTCGTGAGATGTTGGGTTAAGTCCCGC
AACGAGCGCAACCCTTGTCGTTAGTTACTACCATTAAGTTGAGGACTCTAGCGAGACTGCCTGGGTTAAC
CAGGAGGAAGGTGGGGATGACGTCAAATCATCATGCCCCTTATGTGTAGGGCTACACACGTGCTACAATG
GCTGGTACAGAGAGATGCAATACCGCGAGGTGGAGCCAAACTTAAAAACCAGTCTCAGTTCGGATTGTAG
GCTGAAACTCGCCTACATGAAGCTGGAGTTACTAGTAATCGCGAATCAGAATGTCGCGGTGAATACGTTC
CCGGGTCTTGTACACACCGCCCGTCACACCATGAGAGTTGGCAATACCCGAAGTCCGTGAGCTAACCGCA
AGGAGGCAGCGGCCGAAGGTAGGGTCAGCGATTGGGGTG")

# >NR_036786.1:18-1456 Clostridium botulinum strain ELTDK 103 16S ribosomal RNA, partial sequence
ex.Cbot <- gsub("\n", "", "TCAGGACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAGCGATGAAGCTTCCTTCGGGAAGTGGAT
TAGCGGCGGACGGGTGAGTAACACGTGGGTAACCTGCCTCAAAGTGGGGGATAGCCTTCCGAAAGGAAGA
TTAATACCGCATAACATAAGAGAATCGCATGATTTTCTTATCAAAGATTTATTGCTTTGAGATGGACCCG
CGGCGCATTAGCTAGTTGGTAAGGTAACGGCTTACCAAGGCAACGATGCGTAGCCGACCTGAGAGGGTGA
TCGGCCACATTGGAACTGAGACACGGTCCAGACTCCTACGGGAGGCAGCAGTGGGGAATATTGCGCAATG
GGGGAAACCCTGACGCAGCAACGCCGCGTGGGTGATGAAGGTCTTCGGATTGTAAAGCCCTGTTTTCTGG
GACGATAATGACGGTACCAGAGGAGGAAGCCACGGCTAACTACGTGCCAGCAGCCGCGGTAATACGTAGG
TGGCGAGCGTTGTCCGGATTTACTGGGCGTAAAGGGTGCGTAGGCGGATGTTTAAGTGGGATGTGAAATC
CCCGGGCTTAACCTGGGGGCTGCATTCCBAAACTGGATATCTAGAGTGCAGGAGAGGAAAGCGGAATTCC
TAGTGTAGCGGTGAAATGCGTAGAGATTAGGAAGAACACCAGTGGCGAAGGCGGCTTTCTGGACTGTAAC
TGACGCTGAGGCACGAAAGCGTGGGTAGCAAACAGGATTAGATACCCTGGTAGTCCACGCCGTAAACGAT
GGATACTAGGTGTAGGGGGTATCAACTCCCCCTGTGCCGCAGTTAACACAATAAGTATCCCGCCTGGGGA
GTACGGTCGCAAGATTAAAACTCAAAGGAATTGACGGGGGCCCGCACAAGCAGCGGACAGTGTGGTTTAA
TTCGAAGCAACGCAGAGAACCTTACCTGGACTTGACATCCCTTGCATAGCCTAGAGATAGGTGAAGCCCT
TCGGGGCAAGGAGACAGGTGGTGCATGGTTGTCGTCAGCTCGTGTCGTGAGATGTTAGGTTAAGTCCTGC
AACGAGCGCAACCCTTGTTATTAGTTGCTACCATTAAGTTGAGCACTCTAATGAGACTGCCTGGGTAACC
AGGAGGAAGGTGGGGATGACGTCAAATCATCATGCCCCTTATGTCCAGGGCTACACACGTGCTACAATGG
TAGGTACAATAAGACGCAAGACCGTGAGGTGGAGCAAAACTTATAAAACCTATCTCAGTTCGGATTGTAG
GCTGCAACTCGCCTACATGAAGCTGGAGTTGCTAGTAATCGCGAATCAGAATGTCGCGGTGAATACGTTC
CCGGGCCTTGTACACACCGCCCGTCACACCATGAGAGCTGGTAACACCCGAAGTCCGTGAGGTAACCGTA
AGGGAGCCAGCGGCCGAAGTGGATTAAGTGATTGGGGTG")

tt <- assignTaxonomy(c(ex.Cperf, ex.Cbot), "~/tax/silva_nr_v132_train_set.fa.gz", minBoot=80, multi=TRUE)
```

Now taking a look, first is C. perfringens, second is C. botulinum:
```{r}
unname(tt)
sum(tax[,"Genus"] %in% tt[1,"Genus"])
sum(tax[,"Genus"] %in% tt[2,"Genus"])
```

OK so those groups are both in the data, will have to take a closer look at each:

```{r}
sq.Css1 <- getSequences(st)[tax[,"Genus"] %in% "Clostridium_sensu_stricto_1"]
dada2:::pfasta(sq.Css1) # C. perfringens group
```

Yep, there is a lot of *C. perfringens* here.

```{r}
spec.Css1 <- c("perfringens", # 100
"perfringens",
"perfringens",
"perfringens", # 100
"perfringens",
"perfringens",
"perfringens",
"perfringens",
"perfringens",
"septicum", # SEQ 10
"perfringens",
"perfringens",
"septicum",
"septicum",
"perfringens",
"perfringens",
"perfringens", # 23-off
"perfringens",
"perfringens",
"perfringens", # SEQ 20
"septicum",
"perfringens")
names(spec.Css1) <- sq.Css1
```

Now for the C. botulinum group (Css18):

```{r}
sq.Css18 <- getSequences(st)[tax[,"Genus"] %in% "Clostridium_sensu_stricto_18"]
dada2:::pfasta(sq.Css18) # C. botulinum group
```

```{r}
spec.Css18 <- c("sporogenes", # 100
"sporogenes", 
"sporogenes", 
"sporogenes", 
"sporogenes", 
"sporogenes", 
"sporogenes", 
"sporogenes", 
"sporogenes", 
"sporogenes", # SEQ 10
"sporogenes", # 20-off
"sporogenes", # 17-off
"sporogenes", # 16-off, all at end
"sporogenes")
names(spec.Css18) <- sq.Css18
```

No *C. botulinum*.

Now what about Salmonella:
```{r}
sq.salm <- getSequences(tax)[tax[,"Genus"] %in% "Salmonella"]
unname(st[,sq.salm])
```

Interesting, *Salmonella* only shows up in one sample, and in very even counts across 7 ASVs. That is consistent with the 7 rrn copies that *Salmonella* has according to rrndb. So, what is that strain?

```{r}
dada2:::pfasta(sq.salm)
```

Some ASVs (e.g. 1 and 2) are not very discriminative, but Salmonella enterica subsp. enterica serovar Newport strain CFSAN003387, complete genome (https://www.ncbi.nlm.nih.gov/nucleotide/CP016014.1?report=genbank&log$=nucltop&blast_rank=37&RID=9CCPUEXF014) is the best match (usually tied with many others) for every single ASV. All are exact matches except ASV4 (2-off). ASV6 is the key discrimant, as this Newport accession is the only exact match to that ASV. Nice!

STILL TO DO, sq.Css1 (C. perfringens), E. coli, and revisit Yersinia again.

```{r}
sq.Cperf <- names(spec.Css1)[spec.Css1 == "perfringens"]
unname(st[,sq.Cperf])
```

So this pretty much just occurs in the 4th smple. There are singleton detections of other ASVs also i that sample, and in sample 1. Sample 6 has one each of 3/2/1 abundance ASVs. *C. perfringens* has almost all 10 rrn copied in rrndb. This is consistent with the 2:1(x8) ratio of abundance Cperf ASVs in Sample 4. Using the BLAST strategy to try to ID that strain more precisely:

```{r}
dada2:::pfasta(sq.Cperf[1:9])
```

Stores the top hits for each ASV in the `Docs/` subdirectory. Taking a look at the best accessions:
```{r}
hitdf <- read.table("Docs/Food_Sample4_Cperf_BHits.txt", sep="\t", header=TRUE, comment="#")
#hitdf
sort(table(hitdf$Description), decreasing=TRUE)
```

So three strains clearly come out on top w/ 5 best hits each: ATCC 13124, LLY_N11, NCTC13170.

According to https://link.springer.com/article/10.1186/s12864-018-4771-1, ATCC13124 is a non-poultry isolate that lacks pathogenic genes netB and VR-10B. https://www.frontiersin.org/articles/10.3389/fmicb.2017.02485/full also supports that ATCC13124 and NCTC13170 are closely related and lack netB, netE, netF, netG and cpe. (see also comment on this paper https://www.frontiersin.org/articles/10.3389/fmicb.2018.01856/full)

For ATCC13124 (the type strain) see: https://www.ncbi.nlm.nih.gov/pubmed/16825665
NCTC13170: couldn't find much info in the Genbank
LLY_N11: Complete Genome Sequences of Clostridium perfringens Strain LLY_N11, a Pathogenic Isolate of Necrotic Enteritis from a Healthy Chicken

OK, so... not sure where this is at this point in time. In my very brief look at the literature, the phylogenetics/genomics of C. perfringens seems still somewhat underdeveloped, and it is not even clear what the key food-poisoning associated genes are.

Now let's start taking a look at E. coli...

```{r}
sq.ec<- getSequences(tax)[tax[,"Genus"] %in% "Escherichia/Shigella"]
rowSums(st[,sq.ec])
unname(colSums(st[,sq.ec]))
```

Essentially no E. coli in the first 3 samples. Then E. ecoli abundance is dominated by samples 4/6, with 5 having an OOM or 2 fewer reads. A long tail of singletons ASVs. But also a lot of ASVs with significant abundance (e.g. 10+).

```{r}
unname(st[4:6,sq.ec[colSums(st[,sq.ec])>=8]])
```

Sample 4 is full of E. coli, in several different strains. Sample 5 looks like it has a couple low abundance strains of E. coli. Sample 6 has at least a couple relatively abundant E. coli strains. However, separating the ASVs out by strain is not going to be trivial, especially in Sample 4. *E. coli* has 7 rrn genes.

Starting with Sample 6, let's dig further into the abundant strains:
```{r}
unq.ec6 <- getUniques(st[6, sq.ec])
unq.ec6.ab <- unq.ec6[unq.ec6 > 80]
unq.ec6.ab <- sort(unq.ec6.ab, decreasing=TRUE)
dada2:::pfasta(getSequences(unq.ec6.ab), id=paste("Sample6 E. coli ASV, Abundance", unq.ec6.ab))
```

```{r}
hitdf.ec6 <- read.table("Docs/Food_Sample6_AbundantEcoli_BHits.txt", sep="\t", header=TRUE, comment="#", stringsAsFactors=FALSE)
sort(table(hitdf.ec6$Description), decreasing=TRUE)[1:20]
sort(table(hitdf.ec6$Accession), decreasing=TRUE)[1:20]
```

9 hits on that top one? How? Oh... the Description isn't unique, multiple accessions with that Description. Annoying.

```{r}
foo <- hitdf.ec6[!duplicated(hitdf.ec6$Accession),]
desc.map <- foo$Description
names(desc.map) <- foo$Accession
acc <- "CP023349.1"
desc.map["CP023349.1"]
hitdf.ec6$Sequence[hitdf.ec6$Accession == "CP023349.1"]
```

Let's take a look at what ASVs are matching the top accessions (4+ hits and above):
```{r}
bar <- sort(table(hitdf.ec6$Accession), decreasing=TRUE)
accdf <- data.frame(Accession=names(bar), Hits=as.vector(bar), stringsAsFactors = FALSE)
accdf$Description <- desc.map[accdf$Accession]
for(asv in paste0("ASV", seq_along(unq.ec6.ab))) {
  accdf[,asv] <- sapply(accdf$Accession, function(acc) { 
    any(hitdf.ec6$Accession==acc & hitdf.ec6$Sequence==asv) 
  })
}
head(accdf)
```

ASVs 1/4/5/6/9 are the pattern for three of the 5-hit accessions. As a remninder of the abundances here:

```{r}
unname(unq.ec6.ab)[c(1,4,5,6,9)]
```

Not completely obvious how these land on integerish values, but remember also that there may be overlap in some of these ASVs (e.g. ASV1).

```{r}
sapply(colnames(accdf)[c(-1,-2,-3)], function(nm) sum(accdf[,nm]))
table(hitdf.ec6$Sequence)[paste0("ASV", seq_along(unq.ec6.ab))]
```

Interestingly ASVs 8/10/11/12 are particularly unique:
```{r}
accdf[accdf$ASV8 | accdf$ASV10 | accdf$ASV11 | accdf$ASV12,]
```

ASV8 never shows up with any of ASV1-7. The pattern ASV8/11/12 shows up three times. There is also one 8/10/12.

```{r}
unname(unq.ec6.ab)[c(8,11,12)]
unname(unq.ec6.ab)[c(8,10,12)]
```

Not obvious 7-copy integer values, but not out of the question either (e.g. 4:2:1).

Two of those 3-hit accessions (2014C-3338 & 2009C-3133) are Shiga-toxic, but the other two have no tblastn hits for Shiga toxin subunit A (ALL90389.1) or Shiga toxin subunit B (ALL90390.1).

Looking more systematically at the co-occurences between these ASVs as best hits in accessions:
```{r}
asvs <- paste0("ASV", seq_along(unq.ec6.ab))
names(asvs) <- asvs
outer(asvs, asvs, Vectorize(function(x, y) sum(accdf[,x] & accdf[,y])))
```

Woah, ASV1 and ASV9 always co-occur? That's interesting. Heck, could be a 6:1 ASV1/ASV9 type in there. Let's try pulling that component out:

```{r}
foo <- accdf[!accdf$ASV1 & !accdf$ASV9,]
outer(asvs[c(-1, -9)], asvs[c(-1, -9)], Vectorize(function(x, y) sum(foo[,x] & foo[,y])))
```

So, ASV8/10/11/12 are their own component. We'll call that a unique strain:

```{r}
unname(unq.ec6.ab)[c(8,10,11,12)]
```

There's a 3:2:1:1 ratio. 7 alleles.

Pulling that out:

```{r}
exclude <- c("ASV1", "ASV8", "ASV9", "ASV10", "ASV11", "ASV12")
keep <- c(2,3,4,5,6,7)
foo <- accdf[!accdf$ASV1 & !accdf$ASV8 & !accdf$ASV9 & !accdf$ASV10 & !accdf$ASV11 & !accdf$ASV12,]
outer(asvs[keep], asvs[keep], Vectorize(function(x, y) sum(foo[,x] & foo[,y])))
foo <- accdf[!accdf$ASV1 & !accdf$ASV8 & !accdf$ASV9 & !accdf$ASV10 & !accdf$ASV11 & !accdf$ASV12 & accdf$Hits > 1,]
outer(asvs[keep], asvs[keep], Vectorize(function(x, y) sum(foo[,x] & foo[,y])))
v <- unname(unq.ec6.ab)[keep]
v
av <- unname(asvs[keep])
```

I don't know, probably two strains... but how are they mixed together?

```{r}
N <- 2
combos <- combn(6,N)
bar <- t(apply(combos, 2, function(i) 7*v[i]/sum(v[i])))
data.frame(av[combos[1,]], av[combos[2,]], bar)
resid <- apply(bar, 1, function(xx) sum(abs(xx - round(xx))))
plot(resid)
```

```{r}
combos[,order(resid)]
```

Huh I didn't quiet realize how precisely 4:3 that ASV2/ASV3 or ASV2/ASV4 are, as well as ASV5/ASV6.

So... ASV2/3, ASV4/7 and ASV5/6 divide up the counts into pretty close to 4:3 ratios (ASV4/7 the furthest off). But do those double combos occur in BLAST best-hit table? The pairwise combos do, but do they also have other ASVs in them?

```{r}
foo <- accdf[!accdf$ASV1 & !accdf$ASV8 & !accdf$ASV9 & !accdf$ASV10 & !accdf$ASV11 & !accdf$ASV12,]
comat <- 1*as.matrix(foo[,av])
comat[!duplicated(comat),]
```

No, only ASV2/3 exist amonst those two-ASV combos.

```{r}
sq.ec6.ab <- getSequences(unq.ec6.ab); names(sq.ec6.ab) <- asvs
outer(sq.ec6.ab, sq.ec6.ab, nwhamming, vec=TRUE)
outer(sq.ec6.ab[keep], sq.ec6.ab[keep], nwhamming, vec=TRUE)
```

I don't know, I'm just not feeling that confident of anything other than the 8/10/11/12 that is so well supported by co-occurrence on accessions (although the full 4 never shows up), the 3:2:1:1 abundance ratio, and their internal seq similarity and external sequence divergence with the rest of the E. coli sequences.

## Pub figures

First let's plot the relative abundances and identities of the CDC key foodborne pathogens in these 6 samples: Yersinia enterolitica, E. coli (Escherichia/Shigella), Salmonella enterica, Clostridium perfringens (Clostridium_sensu_stricto_1), Campylobacter and Listeria monocytogenes.
```{r}
tt <- tax
tt <- cbind(tt, Species=NA)
tt[sq.yent,"Species"] <- "Y. enterolitica"
tt[sq.Cperf,"Species"] <- "C. perfringens"
tt[sq.salm,"Species"] <- "S. enterica"
tt[sq.ec,"Species"] <- "E. coli"
sum(tt[,"Genus"] %in% "Campylobacter")
sum(tt[,"Genus"] %in% "Listeria")
table(tt[,"Species"], useNA="ifany")
```

OK, so we have our 6 species, only 4 of which are present. Let's make a relative abundance barplot showing their relative abundances in these samples:
```{r}
ft <- sweep(st, 1, rowSums(st), "/")
#is.path <- tt[,"Species"] %in% c("Y. enterolitica", "C. perfringens", "S. enterica", "E. coli", "L. monocytogenes", "Campylobacter sp.")
#table(is.path)
spec.path <- c("Y. enterolitica", "C. perfringens", "S. enterica", "E. coli", "L. monocytogenes", "Campylobacter sp.")
suppressWarnings(ftp <- colsum(ft, group=tt[,"Species"])) # Throws a warning because of the NAs in species, this is OK
ftp <- ftp[,colnames(ftp) %in% spec.path]
ftp
```

```{r}
dfp <- cbind(data.frame(Sample=substr(rownames(ftp), 1, 3), stringsAsFactors=FALSE), 
             data.frame(ftp, stringsAsFactors=FALSE))
dfpm <- melt(dfp, id.vars="Sample", variable.name="Species", value.name="Proportion") ### WARNING: Auto-converts the variable name to a FACTOR
dfpm$Species <- as.character(dfpm$Species)
dfpm$Species <- gsub("[.]{2}", ". ", dfpm$Species)
dfpm$Species <- factor(dfpm$Species, levels=spec.path)
#dfpm$Species <- tt[dfpm$Sequence,"Species"]
table(dfpm$Species, useNA="ifany")
```

```{r}
fill.scale.path <- scale_fill_manual(values=c("C. perfringens"="#1b9e77", "E. coli"="#d95f02", "S. enterica"="#7570b3", 
                                              "Y. enterolitica"="#e7298a", "L. monocytogenes"="tomato4", "Campylobacter sp."="#e6ab02"),
                                     name="Pathogenic\nspecies", drop=FALSE)
ppp <- ggplot(data=dfpm, aes(x=Sample, y=Proportion, fill=Species)) + geom_col(width=0.4) + ylim(0,1) + 
  fill.scale.path + theme_bw() + theme(panel.grid=element_blank())
ppp
ggsave(file.path(path.fig, "Meat_PathogenProportions.pdf"), ppp, width=5, height=3, units="in", useDingbats=FALSE)
ggsave(file.path(path.fig, "Meat_PathogenProportions.png"), ppp, width=5, height=3, units="in")
```

Let's return to C. perfringens as an example of high-res full complement "ribotyping" that is possible with this method. Recall the distribution of C. perfringens ASVs:
```{r}
unname(st[,sq.Cperf])
```

We are interested ine visualizing the "ribotype" of the strain in Sample 4 that spans Cperf ASVs 1-9, with a 2:1(x8) ratio of alleles consistent with the 10 copies of the rrn in Cperf genomes.

```{r}
library(DECIPHER)
library(ggmsa) # On CRAN, also needs BioC treeio package
sq.Cpstrain <- sq.Cperf[1:9]
dna.Cpstrain <- DNAStringSet(sq.Cpstrain)
rna.Cpstrain <- DNAStringSet(dna.Cpstrain)
aln <- AlignSeqs(rna.Cpstrain) # align with RNA secondary structure
names(aln) <- paste0("Cperf", seq_along(sq.Cpstrain))
aln
```

```{r}
## Exxample ggmsa code o pget things working
#library(seqmagick)
#ggmsa(sequences, 320, 360, font = NULL, color = "Clustal")
```

Testing ggmsa nucleotide color schemes:
```{r}
clrs <- c('Clustal', 'Chemistry_NT', 'Shapely_NT', 'Zappo_NT', 'Taylor_NT')
for(clr in clrs) {
  print(ggmsa(aln, 100, 200, font = NULL, color = clr))
}
```

Scannings 100nt chunks to see where some diffs are for the publication plot:
```{r, warnings=FALSE, message=FALSE}
lst <- min(nchar(aln))
for(start in seq(1, 1500, 100)) {
  print(ggmsa(aln, start, min(lst, start+99), font = NULL, color = 'Shapely_NT'))
}
```

Almost all the diffs are in V1/V2 it seems (about postiions 50-75 and 150-190 in these sequences, which start at ~E. coli coordinate position 27, which overlap the V1 and V2 E. coli positions according to https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2562909/).

Maybe just plot V2 using those coords:

```{r}
pmsa <- ggmsa(aln, 137-27, 242-27, font = NULL, color = 'Shapely_NT') + ggtitle("C. perfringen alleles, V2 hypervariable region")
pmsa
ggsave(file.path(path.fig, "Meat_CPerf_MSA.pdf"), pmsa, width=8, height=1.2, units="in", useDingbats=FALSE)
ggsave(file.path(path.fig, "Meat_CPerf_MSA.png"), pmsa, width=8, height=1.2, units="in")
```



